From ec06b1e05a86ac0ce239a46cf98dc324fb3dc012 Mon Sep 17 00:00:00 2001
From: Mohan Kumar G <quic_mkumarg@quicinc.com>
Date: Tue, 27 Aug 2024 08:50:15 +0530
Subject: [PATCH] hostapd: Use BSS specific WMM parameters

Currently the WMM parameters (hostapd_wmm_ac_params) is common for
all BSSes. Hence if any of the BSS WMM parameters is changed, it is
getting changed for all the other BSSes also.

By default WMM parameter is initialized as an interface config and is
shared by all BSSes in the interface. When WMM parameter of specific
BSS is changed via hostapd_cli command, it is updated in the
interface config and gets changed for all other BSSes.

Hence, initialize a WMM parameter in the BSS config and whenever BSS
specific change is applied, update in this BSS parameter instead of
interface parameter. Also use wmm_override flag to override any
interface specific WMM parameter with BSS specific one if BSS is
specified.

When changing WMM parameter, MU EDCA gets updated for all BSSes.
Change MU EDCA to update only for the BSSes where WMM change in
applied.

ToDo: Add MU EDCA IE for non-Tx BSS in MBSSID profile in
hostapd_eid_mbssid_elem() function.

Patch-depdendency: r35-002-hostapd-set-critical-update-for-non-transmitting-BSS.patch

Patch-work:

Feature: MBSSID

Co-developed-by: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Signed-off-by: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Signed-off-by: Mohan Kumar G <quic_mkumarg@quicinc.com>
---
 hostapd/config_file.c  |  4 +++-
 hostapd/ctrl_iface.c   |  6 ------
 src/ap/ap_config.c     | 26 ++++++++++++++++++++++++++
 src/ap/ap_config.h     |  3 +++
 src/ap/drv_callbacks.c |  7 +------
 src/ap/hostapd.h       |  2 ++
 src/ap/ieee802_11_he.c |  9 ++++++++-
 src/ap/wmm.c           | 18 ++++++++++--------
 8 files changed, 53 insertions(+), 22 deletions(-)

diff --git a/hostapd/config_file.c b/hostapd/config_file.c
index daefd34..92850cf 100644
--- a/hostapd/config_file.c
+++ b/hostapd/config_file.c
@@ -3584,9 +3584,11 @@ static int hostapd_config_fill(struct hostapd_config *conf,
 		bss->wmm_uapsd = atoi(pos);
 	} else if (os_strncmp(buf, "wme_ac_", 7) == 0 ||
 		   os_strncmp(buf, "wmm_ac_", 7) == 0) {
-		if (hostapd_config_wmm_ac(conf->wmm_ac_params, buf, pos)) {
+		bss->wmm_override = true;
+		if (hostapd_config_wmm_ac(bss->wmm_ac_params, buf, pos)) {
 			wpa_printf(MSG_ERROR, "Line %d: invalid WMM ac item",
 				   line);
+			bss->wmm_override = false;
 			return 1;
 		}
 	} else if (os_strcmp(buf, "bss") == 0) {
diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 1d5c2cc..6f386bb 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -1322,12 +1322,6 @@ static int hostapd_ctrl_iface_set(struct hostapd_data *hapd, char *cmd)
 		} else if (os_strncmp(cmd, "wme_ac_", 7) == 0 ||
 			   os_strncmp(cmd, "wmm_ac_", 7) == 0) {
 			hapd->parameter_set_count++;
-			 /* Incrementing MU-EDCA Parameter Set Update Count*/
-			 hapd->iface->conf->he_mu_edca.he_qos_info =
-			  (hapd->iface->conf->he_mu_edca.he_qos_info & 0xf0) |
-			  ((hapd->iface->conf->he_mu_edca.he_qos_info + 1) &
-			   0xf);
-
 			if (tx_hapd && ieee802_11_update_beacons(tx_hapd->iface))
 				wpa_printf(MSG_DEBUG,
 					   "Failed to update beacons with WMM parameters");
diff --git a/src/ap/ap_config.c b/src/ap/ap_config.c
index 12bdad5..d5b2c4b 100644
--- a/src/ap/ap_config.c
+++ b/src/ap/ap_config.c
@@ -176,6 +176,32 @@ void hostapd_config_defaults_bss(struct hostapd_bss_config *bss)
 	bss->pasn_comeback_after = 10;
 	bss->pasn_noauth = 1;
 #endif /* CONFIG_PASN */
+
+	bss->wmm_override = false;
+#define ecw2cw(ecw) ((1 << (ecw)) - 1)
+	const int aCWmin = 4, aCWmax = 10;
+	const struct hostapd_wmm_ac_params ac_bk =
+		{ aCWmin, aCWmax, 7, 0, 0 }; /* background traffic */
+	const struct hostapd_wmm_ac_params ac_be =
+		{ aCWmin, aCWmax, 3, 0, 0 }; /* best effort traffic */
+	const struct hostapd_wmm_ac_params ac_vi = /* video traffic */
+		{ aCWmin - 1, aCWmin, 2, 3008 / 32, 0 };
+	const struct hostapd_wmm_ac_params ac_vo = /* voice traffic */
+		{ aCWmin - 2, aCWmin - 1, 2, 1504 / 32, 0 };
+	const struct hostapd_tx_queue_params txq_bk =
+		{ 7, ecw2cw(aCWmin), ecw2cw(aCWmax), 0 };
+	const struct hostapd_tx_queue_params txq_be =
+		{ 3, ecw2cw(aCWmin), 4 * (ecw2cw(aCWmin) + 1) - 1, 0};
+	const struct hostapd_tx_queue_params txq_vi =
+		{ 1, (ecw2cw(aCWmin) + 1) / 2 - 1, ecw2cw(aCWmin), 30};
+	const struct hostapd_tx_queue_params txq_vo =
+		{ 1, (ecw2cw(aCWmin) + 1) / 4 - 1,
+		  (ecw2cw(aCWmin) + 1) / 2 - 1, 15};
+#undef ecw2cw
+	bss->wmm_ac_params[0] = ac_be;
+	bss->wmm_ac_params[1] = ac_bk;
+	bss->wmm_ac_params[2] = ac_vi;
+	bss->wmm_ac_params[3] = ac_vo;
 }
 
 
diff --git a/src/ap/ap_config.h b/src/ap/ap_config.h
index 559fe41..6e4bb69 100644
--- a/src/ap/ap_config.h
+++ b/src/ap/ap_config.h
@@ -988,6 +988,9 @@ struct hostapd_bss_config {
 	bool mld_indicate_disabled;
 #endif /* CONFIG_TESTING_OPTIONS */
 #endif /* CONFIG_IEEE80211BE */
+
+	bool wmm_override;
+	struct hostapd_wmm_ac_params wmm_ac_params[4];
 };
 
 /**
diff --git a/src/ap/drv_callbacks.c b/src/ap/drv_callbacks.c
index 83a142e..e941dde 100644
--- a/src/ap/drv_callbacks.c
+++ b/src/ap/drv_callbacks.c
@@ -2352,7 +2352,6 @@ static void hostapd_event_update_muedca_params(struct hostapd_data *hapd,
 					       struct update_muedca *params)
 {
 	int i;
-	u8 updated_count;
 
 	/* Update current MU-EDCA parameters */
 	for (i = 0; i < 3; i++) {
@@ -2369,12 +2368,8 @@ static void hostapd_event_update_muedca_params(struct hostapd_data *hapd,
 	/* Increment Parameter Set Update Count for MU-EDCA and WME EDCA only
 	 * if any STA is connected
 	 */
-	if (hapd->num_sta) {
-		updated_count = (hapd->iface->conf->he_mu_edca.he_qos_info + 1) & 0xf;
-		hapd->iface->conf->he_mu_edca.he_qos_info &= 0xf0;
-		hapd->iface->conf->he_mu_edca.he_qos_info |= updated_count;
+	if (hapd->num_sta)
 		hapd->parameter_set_count++;
-	}
 
 	/* Update beacon with updated MU-EDCA parameters */
 	if (ieee802_11_update_beacons(hapd->iface))
diff --git a/src/ap/hostapd.h b/src/ap/hostapd.h
index 405c3bd..0690da5 100644
--- a/src/ap/hostapd.h
+++ b/src/ap/hostapd.h
@@ -550,6 +550,8 @@ struct hostapd_data {
 	u64 scan_cookie; /* Scan instance identifier for the ongoing HT40 scan
 			  */
 	bool disable_cu;
+
+	struct hostapd_wmm_ac_params prev_wmm[4];
 };
 
 
diff --git a/src/ap/ieee802_11_he.c b/src/ap/ieee802_11_he.c
index 3fc087e..97d0850 100644
--- a/src/ap/ieee802_11_he.c
+++ b/src/ap/ieee802_11_he.c
@@ -306,7 +306,7 @@ u8 * hostapd_eid_he_mu_edca_parameter_set(struct hostapd_data *hapd, u8 *eid)
 {
 	struct ieee80211_he_mu_edca_parameter_set *edca;
 	struct hostapd_wmm_ac_params wmmp[WMM_AC_NUM];
-	u8 *pos;
+	u8 *pos, updated_count;
 	size_t i;
 
 	 /* Updating WME Parameter Set Count to avoid mismatch */
@@ -331,6 +331,13 @@ u8 * hostapd_eid_he_mu_edca_parameter_set(struct hostapd_data *hapd, u8 *eid)
 	edca = (struct ieee80211_he_mu_edca_parameter_set *) pos;
 	os_memcpy(edca, &hapd->iface->conf->he_mu_edca, sizeof(*edca));
 
+	updated_count = edca->he_qos_info & 0xf;
+	if (updated_count != hapd->parameter_set_count & 0xf) {
+		updated_count = hapd->parameter_set_count & 0xf;
+		edca->he_qos_info &= 0xf0;
+		edca->he_qos_info |= updated_count;
+	}
+
 	wpa_hexdump(MSG_DEBUG, "HE: MU EDCA Parameter Set element",
 		    pos, sizeof(*edca));
 
diff --git a/src/ap/wmm.c b/src/ap/wmm.c
index a3fe1e6..6387e06 100644
--- a/src/ap/wmm.c
+++ b/src/ap/wmm.c
@@ -66,8 +66,14 @@ void wmm_calc_regulatory_limit(struct hostapd_data *hapd,
 {
 	struct hostapd_hw_modes *mode = hapd->iface->current_mode;
 	int c;
+	struct hostapd_wmm_ac_params *wmm_ac_params = hapd->iconf->wmm_ac_params;
+	struct hostapd_wmm_ac_params *prev_wmm = hapd->iface->prev_wmm;
 
-	os_memcpy(acp, hapd->iconf->wmm_ac_params,
+	if (hapd->conf->wmm_override) {
+		wmm_ac_params = hapd->conf->wmm_ac_params;
+		prev_wmm = hapd->prev_wmm;
+	}
+	os_memcpy(acp, wmm_ac_params,
 		  sizeof(hapd->iconf->wmm_ac_params));
 
 	for (c = 0; mode && c < mode->num_channels; c++) {
@@ -77,7 +83,7 @@ void wmm_calc_regulatory_limit(struct hostapd_data *hapd,
 			continue;
 
 		if (chan->wmm_rules_valid)
-			wmm_set_regulatory_limit(hapd->iconf->wmm_ac_params,
+			wmm_set_regulatory_limit(wmm_ac_params,
 						 acp, chan->wmm_rules);
 		break;
 	}
@@ -86,15 +92,11 @@ void wmm_calc_regulatory_limit(struct hostapd_data *hapd,
 	 * Check if we need to update set count. Since both were initialized to
 	 * zero we can compare the whole array in one shot.
 	 */
-	if (os_memcmp(acp, hapd->iface->prev_wmm,
+	if (os_memcmp(acp, prev_wmm,
 		      sizeof(hapd->iconf->wmm_ac_params)) != 0) {
-		os_memcpy(hapd->iface->prev_wmm, acp,
+		os_memcpy(prev_wmm, acp,
 			  sizeof(hapd->iconf->wmm_ac_params));
 		hapd->parameter_set_count++;
-		 /* Incrementing MU-EDCA Parameter Set Update Count*/
-		  hapd->iface->conf->he_mu_edca.he_qos_info =
-		  (hapd->iface->conf->he_mu_edca.he_qos_info & 0xf0) |
-		  ((hapd->iface->conf->he_mu_edca.he_qos_info + 1) & 0xf);
 	}
 }
 
-- 
2.34.1

